crates/turborepo-lib/src/run/watch.rs

```diff
@@ -76,43 +109,58 @@ impl WatchClient {
             execution_args: execution_args.clone(),
         });
 
-        let mut run = RunBuilder::new(new_base)?
+        let run = RunBuilder::new(new_base)?
             .build(&handler, telemetry.clone())
             .await?;
 
-        run.print_run_prelude();
-
         let connector = DaemonConnector {
             can_start_server: true,
             can_kill_server: true,
             paths: DaemonPaths::from_repo_root(&base.repo_root),
         };
 
+        Ok(Self {
+            base,
+            run,
+            connector,
+            handler,
+            telemetry,
+            persistent_tasks_handle: None,
+        })
+    }
+
+    pub async fn start(&mut self) -> Result<(), Error> {
+        let connector = self.connector.clone();
         let mut client = connector.connect().await?;
 
         let mut events = client.package_changes().await?;
-        let mut current_runs: HashMap<PackageName, JoinHandle<Result<i32, run::Error>>> =
-            HashMap::new();
-        let mut persistent_tasks_handle = None;
+
+        self.run.print_run_prelude();
+
+        let signal_subscriber = self.handler.subscribe().ok_or(Error::NoSignalHandler)?;
+
+        let (changed_pkgs_tx, mut changed_pkgs_rx) = watch::channel(ChangedPackages::default());
 
         let event_fut = async {
             while let Some(event) = events.next().await {
                 let event = event?;
-                Self::handle_change_event(
-                    &mut run,
-                    event.event.unwrap(),
-                    &mut current_runs,
-                    &base,
-                    &telemetry,
-                    &handler,
-                    &mut persistent_tasks_handle,
-                )
-                .await?;
+                Self::handle_change_event(&changed_pkgs_tx, event.event.unwrap()).await?;
             }
 
             Err(Error::ConnectionClosed)
         };
 
+        let run_fut = async {
+            loop {
+                changed_pkgs_rx.changed().await?;
+                let changed_pkgs = changed_pkgs_rx.borrow_and_update();
```

Comment:
I think this will block the producing side until `changed_pkgs` is dropped. You may want to clone it so that the borrow is short-lived, or at least not held for the duration of a run.

Reply:
Docs on borrowing [here](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.borrow_and_update) indicate that this is not recommended, and we should find a way to shorten the borrow.